rockchip: add support for Orange Pi R1 Plus

nebulakl opened the pull request:
This PR adds support for the Orange Pi R1 Plus, a single board computer based on Rockchip RK3328 SoC.

Specification:
- CPU: Rockchip RK3328 @ 1.5 GHz
- RAM: 1GB DDR4
- Storage: microSD, 16MB SPI Flash
- Network:
  - 10M/100M/1000M Ethernet (RTL8211E)
  - 10M/100M/1000M USB Ethernet (RTL8153B)
- LED: WAN, LAN, Status
- USB: 1x USB 2.0 HOST
- Button: 1x reset button
- Debug UART: 3v3 @ 1500000 baud rate

Flash instruction:
- Write sysupgrade image to a microSD card using dd
- Insert the card, and power on the board

Notes on MAC address:
The USB Ethernet (LAN) has a stable, unique MAC address.
MAC address for WAN is LAN's address minus 1.

Signed-off-by: Honpan Lung <lunghonpan@gmail.com>

adschm commented on target/linux/rockchip/armv8/base-files/etc/board.d/01_leds:
```diff
@@ -8,7 +8,7 @@ boardname="${board##*,}"
 
 case $board in
-friendlyarm,nanopi-r2s)
+friendlyarm,nanopi-r2s|xunlong,orangepi-r1-plus)
```
Please use multiline statements here (also below):

```
friendlyarm,nanopi-r2s|\
xunlong,orangepi-r1-plus)
```

adschm replied:
After boardname is dropped in DTS, this should be adjusted accordingly.

adschm commented on target/linux/rockchip/armv8/base-files/etc/board.d/02_network:
```diff
@@ -35,6 +35,10 @@ rockchip_setup_macs()
 		lan_mac=$(macaddr_add "$wan_mac" +1)
 		;;
+	xunlong,orangepi-r1-plus)
+		lan_mac=$(cat /sys/class/net/eth1/address)
```
Is this necessary? Shouldn't lan be eth1 anyway?

adschm commented on target/linux/rockchip/armv8/base-files/etc/hotplug.d/net/40-net-smp-affinity:
```diff
@@ -22,7 +22,7 @@ set_interface_core() {
 
 case "$(board_name)" in
-friendlyarm,nanopi-r2s)
+friendlyarm,nanopi-r2s|xunlong,orangepi-r1-plus)
```
See above.

adschm commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -12,6 +12,16 @@ define Device/friendlyarm_nanopi-r2s
 endef
 TARGET_DEVICES += friendlyarm_nanopi-r2s
 
+define Device/xunlong_orangepi-r1-plus
```
Please put this block at the correct position according to alphabetic sorting.

adschm commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -12,6 +12,16 @@ define Device/friendlyarm_nanopi-r2s
+  DEVICE_MODEL := OrangePi R1 Plus
+  SOC := rk3328
+  UBOOT_DEVICE_NAME := orangepi-r1-plus-rk3328
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script orangepi-r1-plus | pine64-img | gzip | append-metadata
```
Please wrap lines after 74 chars.

adschm commented on target/linux/rockchip/patches-5.10/106-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,366 @@
++
++		wan_led: led-0 {
++			gpios = <&gpio2 RK_PC2 GPIO_ACTIVE_HIGH>;
++			label = "orangepi-r1-plus:green:wan";
```
Please drop the model here, label should be just "color:function", i.e. green:wan etc.

nebulakl replied:
I intend to submit this patch to upstream Linux. Is it still necessary to make such a change because from [LED handling under Linux](https://www.kernel.org/doc/html/latest/leds/leds-class.html):
> LED Device Naming is currently of the form: `“devicename:color:function”`

adschm replied:
Please read the entire section.

Devicename refers to internal units, not to the "model name". Since I don't see a relevant device to refer to here, omitting devicename seems the right option here.

BTW the whole label for LEDs is deprecated in kernel ...

adschm replied:
And if you really read the whole section you will also find examples ...

nebulakl replied:
Ok cool. I have corrected the device tree file, so please take a look.

adschm commented on target/linux/rockchip/patches-5.4/105-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,362 @@
++
++		wan_led: led-0 {
++			gpios = <&gpio2 RK_PC2 GPIO_ACTIVE_HIGH>;
++			label = "orangepi-r1-plus:green:wan";
```
See above.

hoatienii commented:
how to add support for Orange Pi R1 Plus. I'm newbie, thanks so much!

ronnsn commented:
How can we get this missing 2 approving reviewers?

necro-nemesis commented:
> How can we get this missing 2 approving reviewers?

Any possibility of getting this 2 week old request approved to merge?

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-plat.c:
```diff
@@ -0,0 +1,153 @@
+/*
```
Please add an SPDX license identifier for all new files (except *.patch).

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-plat.c:
```diff
@@ -0,0 +1,153 @@
+	.plat_size	= sizeof(dtv_syscon_at_ff100000),
+	.parent_idx	= -1,
+};
+
```
remove empty line at EOF.

ronnsn replied:
@nebulakl would you pls be kind to remove the empty line, then @adschm can close this. Thx.

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-structs-gen.h:
```diff
@@ -0,0 +1,59 @@
+/*
```
SPDX, see above ...

adschm commented on target/linux/rockchip/armv8/base-files/etc/board.d/02_network:
```diff
@@ -35,6 +36,9 @@ rockchip_setup_macs()
 		lan_mac=$(macaddr_add "$wan_mac" +1)
 		;;
+	xunlong,orangepi-r1-plus)
+		wan_mac=$(macaddr_add "$(cat /sys/class/net/eth1/address)" -1)
```
Can the base address be read from flash somehow? Having relative definitions for lan/wan is somewhat unpleasant when things change later ...

nebulakl replied:
Unfortunately no. If I understand correctly this is how the [vendor firmware](https://github.com/orangepi-xunlong/orangepi-build/blob/004741d87c3ad6022849a553b6c75aa2c199c2f5/external/packages/blobs/rockchip/set_r1plus_lan_mac.sh) sets the MAC address.
There are references to EEPROM with MAC address in the schematic, but it also says this part is NC (no connect).

adschm commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -30,3 +30,14 @@ define Device/radxa_rock-pi-4
+  SOC := rk3328
+  UBOOT_DEVICE_NAME := orangepi-r1-plus-rk3328
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script nanopi-r2s |\
+                             pine64-img | gzip | append-metadata
```
Please use one single tab for hanging indent.

adschm commented on target/linux/rockchip/patches-5.10/106-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,368 @@
++#include "rk3328.dtsi"
++
++/ {
++	model = "Orange Pi R1 Plus";
```
"OrangePi" or "Orange Pi". Please consistently use one option (also for DEVICE_MODEL and patches-5.4. Based on your compatible/device name, the variant without space should be used.

nebulakl replied:
Let's keep it as "Orange Pi" because the devices made by Xunlong/Orange Pi are named this way in the upstream linux.

adschm replied:
Okay, if the kernel is consistently inconsistent, we should probably follow this scheme.

adschm commented on target/linux/rockchip/patches-5.10/106-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,368 @@
+--- /dev/null
++++ b/arch/arm64/boot/dts/rockchip/rk3328-orangepi-r1-plus.dts
+@@ -0,0 +1,355 @@
++// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
```
"GPL-2.0-or-later OR MIT"
same for 5.4

ronnsn commented:
@nebulakl -  thank you so much for your efforts and your great work!! 
Question:  does the config.buildinfo file in https://downloads.openwrt.org/snapshots/targets/rockchip/armv8/   also need an update, so that the snapshot images can be built automatically from the system?

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-plat.c:
```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: GPL-2.0+
```
"GPL-2.0-or-later

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-plat.c:
```diff
@@ -0,0 +1,153 @@
+	.reg			= {0xff400000, 0x1000, 0xff780000, 0x3000, 0xff100000, 0x1000, 0xff440000, 0x1000,
+		0xff720000, 0x1000, 0xff798000, 0x1000},
+	.rockchip_sdram_params	= {0x1, 0xa, 0x2, 0x1, 0x0, 0x0, 0x11, 0x0,
+		0x11, 0x0, 0x0, 0x94291288, 0x0, 0x27, 0x462, 0x15,
```
This indenting scheme looks strange ...

nebulakl replied:
Is the style of this file and `dt-structs-gen.h` important? 
Normally these two files are generated during the compilation of U-boot and are only included here to work around the missing dependency problem in OpenWrt.

Source: https://github.com/openwrt/openwrt/blob/946019637ec9192450c685a61c589a9cbef7637b/package/boot/uboot-rockchip/patches/002-spl-remove-dtoc-of-pdata-generation.patch

ronnsn replied:
@adschm is this still an issue or can this be closed?

adschm commented on package/boot/uboot-rockchip/src/of-platdata/orangepi-r1-plus-rk3328/dt-structs-gen.h:
```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-2.0+
```
GPL-2.0-or-later
also below if applicable

adschm commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -41,3 +41,14 @@ define Device/radxa_rock-pi-4
+    pine64-img | gzip | append-metadata
+  DEVICE_PACKAGES := kmod-usb-net-rtl8152 r8169-firmware
+endef
+TARGET_DEVICES += xunlong_orangepi-r1-plus
```
Missing newline at eof.

adschm commented on target/linux/rockchip/image/armv8.mk:
```diff
@@ -41,3 +41,14 @@ define Device/radxa_rock-pi-4
+  SOC := rk3328
+  UBOOT_DEVICE_NAME := orangepi-r1-plus-rk3328
+  IMAGE/sysupgrade.img.gz := boot-common | boot-script nanopi-r2s |\
+    pine64-img | gzip | append-metadata
```
Indent should be one single tab in this file.

teqlee commented:
everything seem ok, why it's still not merge ?

necro-nemesis commented:
Yes I'd like to see this one officially added. Board handles OpenWrt well and is an asset worth supporting.

necro-nemesis commented:
Can we get this reviewed before it goes stale please?

necro-nemesis reviewed (approved):
Test builds work properly

admiral-bokoblin commented:
30 commits behind master

necro-nemesis commented:
> 30 commits behind master

Still awaiting approval to see it merged into master. I don't know what it requires to get that done. Maybe a PR into the latest branch? Not sure what the hold up is at this point.

admiral-bokoblin commented:
> Still awaiting approval to see it merged into master. I don't know what it requires to get that done. Maybe a PR into the latest branch? Not sure what the hold up is at this point.

¯\_(ツ)_/¯ ive been using scripts daily to merge with upstream, compile. and flash the sysupgrade. Not the ideal setup, but itll work until someone approves this so i can freeze on a commit that works well enough until a stable release comes out...

nebulakl commented:
> > Still awaiting approval to see it merged into master. I don't know what it requires to get that done. Maybe a PR into the latest branch? Not sure what the hold up is at this point.
> 
> ¯_(ツ)_/¯ ive been using scripts daily to merge with upstream, compile. and flash the sysupgrade. Not the ideal setup, but itll work until someone approves this so i can freeze on a commit that works well enough until a stable release comes out...

That is very nice of you. Please lmk asap if there are any changes made by upstream break the compilation.

I don't see the necessity to keep my fork synced daily because if you look at the commits, oftentimes they are irrelevant (e.g. QoL updates for other hardware platforms). However, I will update my fork if 1. someone commented on the quality of this PR; 2. I want to make changes to my existing setup; 3. there are changes to `package/boot/uboot-rockchip` and/or `target/linux/rockchip`.

necro-nemesis commented:
> > > Still awaiting approval to see it merged into master. I don't know what it requires to get that done. Maybe a PR into the latest branch? Not sure what the hold up is at this point.
> > 
> > 
> > ¯_(ツ)_/¯ ive been using scripts daily to merge with upstream, compile. and flash the sysupgrade. Not the ideal setup, but itll work until someone approves this so i can freeze on a commit that works well enough until a stable release comes out...
> 
> That is very nice of you. Please lmk asap if there are any changes made by upstream break the compilation.
> 
> I don't see the necessity to keep my fork synced daily because if you look at the commits, oftentimes they are irrelevant (e.g. QoL updates for other hardware platforms). However, I will update my fork if 1. someone commented on the quality of this PR; 2. I want to make changes to my existing setup; 3. there are changes to `package/boot/uboot-rockchip` and/or `target/linux/rockchip`.

I agree with your assessment to update. Just wondering if a PR as well to the latest branch might elevate the visibility again. I'm not sure why it hasn't gotten assigned for approving review. It appeared that what was requested was reasonably promptly addressed by you and it seemed to stall there.

admiral-bokoblin commented:
> Please lmk asap if there are any changes made by upstream break the compilation.
Everythings fine and likely will be for the foreseeable future. Im just a dummy who learned how to use git :P. THANK YOU for all the awesome work and hopefully this gets upstreamed soon!

ronnsn commented:
Please apologize me, if my post is total nonsense - i am not familiar with the openwrt development process.

>I'm not sure why it hasn't gotten assigned for approving review. 

@nebulakl or @adschm have you tried setting the "Needs review" label to push forward`?

For reference, see the overview of existing labels and especially portings with "Needs review" label:
https://github.com/openwrt/openwrt/labels?page=1&sort=name-asc

Many thanks!

adschm commented:
I did put "needs reviewer" here, although I don't think this will make any difference.

necro-nemesis commented:
It might if they are using "needs reviewer" to filter open request searches. Might have to start getting "vocal" on OpenWrt forum and open a topic to add Pi R1 Plus support there then link this PR to it. A PR seemed like the correct process but maybe there's some other?

adschm commented:
> It might if they are using "needs reviewer" to filter open request searches.

I don't really think anybody does that. It's more to signal that it's not the submitters fault that something is lying around for a long time.

> Might have to start getting "vocal" on OpenWrt forum and open a topic to add Pi R1 Plus support there then link this PR to it. A PR seemed like the correct process but maybe there's some other?

Yes, a PR is the right way to submit device support. However, some platforms are simply chronically short on reviewers. There is not much you can do about that, except maybe find some people without commit access who will give you Tested-by or Reviewed-by here. Of course, the change of getting a Reviewed-by from a non-committer is not so much bigger given the amount of changes required by the submission.

teqlee commented:
How to become "tested-by" or "reviewed-by" ? I have this board and used this PR without issue.

necro-nemesis commented:
> How to become "tested-by" or "reviewed-by" ? I have this board and used this PR without issue.

You can add your own approving review at the bottom link "Add your review" but it requires 2 approving reviews from reviewers with "write access".  I suppose the more people that comment that it's working well for them substantiates getting these other 2 approvals.

necro-nemesis reviewed (approved):
Reviewed

EvilGremlin commented:
please merge it already, it works

nebulakl commented:
I accidentally force-pushed a wrong commit and closed the PR.

necro-nemesis commented:
> I accidentally force-pushed a wrong commit and closed the PR.

Ah. I've been looking at your and Nano Pi's code and it would seem that if the kernel patches for Rockchip 5.4 can be worked out a build for 21.02 may be possible with some minor changes to the config files. If nothing comes of this what do you think about building for 21.02 and trying to get a merge in there?

EvilGremlin commented:
I compiled 21.02 with just simple copypaste. Had no chance to test if it actually works though, probably later today.

necro-nemesis commented:
Let us know please. This PR is quite buried in a pile of other OpewWrt PRs at this late stage. I'm thinking to get a fresh look at merging it the requirement might be to PR 21.02 at this point.  Subsequently a request can be made to have them circle back to PR the same code into 19.07 for anyone still wanting to use it. It's likely they are now focusing on improvements to 21.02.

EvilGremlin commented:
Yep, it booted and it works, just simple copypaste.

nebulakl commented:
Perhaps @blocktrron would be willing to review this PR and help get it merged.

necro-nemesis commented:
> Yep, it booted and it works, just simple copypaste. Though i have no luci for some reason, and it all looks congusing... I've built this:
> 
> ```
> ~/openwrt$ git branch 
> * (HEAD отделён начиная с 79103d6de2)
>   master
>   v21.02.1
> ```
> 
> but in shell it says ` OpenWrt SNAPSHOT, r18086+2-cb18b62206` and kernel is 5.10.79 lolwut? and opkg targets at snapshot, which obviously doesn't work and signatire is wrong... How do i build 21.02 right? What i've done wrong here? https://github.com/EvilGremlin/openwrt/tree/openwrt-21.02/bin/targets/rockchip/armv8

https://github.com/EvilGremlin/openwrt/blob/openwrt-21.02/target/linux/rockchip/Makefile Your Makefile is instructing it to use kernel patches for 5.1 vice 5.4.

necro-nemesis commented:
Building on 21.02 with kernel version 5.4 fails at building uboot. As far as I see it they've moved away from supporting kernel 5.1 (won't fetch) and now 5.4 is required for 21.02 with the associated kernel patches. This requires changes to the Makefile previously stated. There's also device tree patches that need to be added to uboot as well which are a few 5.4 related files that aren't present. A look at the files in `patches/boot/uboot-rockchip/patches`, `/patches/boot/uboot-rockchip/src/of-platdata`, the $topdir Makefile requirement for dt-plat.c to be dt-platdata.c and `targe/linux/rockchip/armv8/` requiring a config-5.4 at the very least it would appear.

walmartshopper commented:
Not sure if this is helpful, but I just got this device and was able to clone master, apply 4093.patch cleanly, and build successfully. It's running kernel 5.10.80 with version string OpenWrt SNAPSHOT r18179-398cbb76fa.  I did have to manually select the r8169 driver during menuconfig.  Using as daily driver and so far everything is working great.

miguemely commented:
Forum thread: https://forum.openwrt.org/t/orange-pi-r1-plus-support/84824/124

Because there's probably someone like me who was wondering what was going on here (like 4093.patch)

necro-nemesis commented:
> Not sure if this is helpful, but I just got this device and was able to clone master, apply 4093.patch cleanly, and build successfully. It's running kernel 5.10.80 with version string OpenWrt SNAPSHOT r18179-398cbb76fa. I did have to manually select the r8169 driver during menuconfig (should DEVICE_PACKAGES include kmod-r8169 instead of r8169-firmware?) Using as daily driver and so far everything is working great.
> 
> I got some errors when I tried to apply the patch to 21.02.1

Yes it's very helpful in keep this effort alive and moving forward. I linked the instructions you posted for the benefit of others. https://forum.openwrt.org/t/orange-pi-r1-plus-support/84824/124?

nebulakl commented:
> Not sure if this is helpful, but I just got this device and was able to clone master, apply 4093.patch cleanly, and build successfully. It's running kernel 5.10.80 with version string OpenWrt SNAPSHOT r18179-398cbb76fa. I did have to manually select the r8169 driver during menuconfig (should DEVICE_PACKAGES include kmod-r8169 instead of r8169-firmware?) Using as daily driver and so far everything is working great.
> 
> I got some errors when I tried to apply the patch to 21.02.1

I am not sure if `kmod-r8169` and `r8169-firmware` are needed even. 
For `kmod-r8169`, I don't see it is being used based on `dmesg` output. 
For `r8169-firmware`, the kernel complains about not being able to load a firmware file even though it exists.

mj22226 commented:
> I am not sure if `kmod-r8169` and `r8169-firmware` are needed even.

@nebulakl  `r8169-firmware` is only needed for v5.10+ together with `kmod-crypto-sha256 ` or use https://github.com/openwrt/openwrt/pull/4356

walmartshopper commented:
> For `kmod-r8169`, I don't see it is being used based on `dmesg` output.

I think you're right, looks like it's using generic phy / gmac driver:

[   11.628349] rk_gmac-dwmac ff540000.ethernet eth0: PHY [stmmac-0:01] driver [RTL8211E Gigabit Ethernet] (irq=POLL)

nebulakl commented:
added `kmod-crypto-sha256` so the usb eth firmware loads.

lessload commented:
"Orange Pi R1 Plus" seem no more produced. Xunlong produce same product with the name "Orange Pi R1 Plus LTS" and replace RTL8211E chip by YT8531C chip + replace Dual ram to new single channel ram.
ref: http://www.orangepi.org/Orange%20Pi%20R1%20Plus%20LTS/

necro-nemesis commented:
https://github.com/orangepi-xunlong/openwrt/tree/openwrt-21.02 Untested. Seems to support LTS and non LTS boards.

nebulakl commented:
Removed patches for kernel 5.4

jmlzr commented:
What should be done about the LTS version?

I would like to help support it but considering there is significant overlap with the non-LTS I'm not sure where to start. Using an unmerged PR as a base for another PR feels wrong and it looks like this PR is not gonna be merged any time soon. Starting from current master seems wrong as well as it would duplicate work and make eventually merging both PRs even harder...

@nebulakl What do you think? Or would you like to integrate LTS support directly into this PR? I have (only) a LTS board for testing and am willing to help.

nebulakl commented:
> What should be done about the LTS version?
> 
> I would like to help support it but considering there is significant overlap with the non-LTS I'm not sure where to start. Using an unmerged PR as a base for another PR feels wrong and it looks like this PR is not gonna be merged any time soon. Starting from current master seems wrong as well as it would duplicate work and make eventually merging both PRs even harder...
> 
> @nebulakl What do you think? Or would you like to integrate LTS support directly into this PR? I have (only) a LTS board for testing and am willing to help.

I would love to take on the job, but I am currently busy with school, so I can't really do it.

Also, Orange Pi seems to be doing something funny with their schematic diagram (e.g. it says RTL8211F instead of YT8531C for ethernet). Without access to an LTS board, adding support may involve unnecessary guesswork, and it is better to avoid that.

As for PR for the LTS board, maybe we will have a better chance to get it merged if someone creates a patch and sends it via the mailing list.

nebulakl committed:
rockchip: add support for Orange Pi R1 Plus

Specification:
- CPU: Rockchip RK3328 @ 1.5 GHz
- RAM: 1GB DDR4
- Storage: microSD, 16MB SPI Flash
- Network:
  - 10M/100M/1000M Ethernet (RTL8211E)
  - 10M/100M/1000M USB Ethernet (RTL8153B)
- LED: WAN, LAN, Status
- USB: 1x USB 2.0 HOST
- Button: 1x reset button
- Debug UART: 3v3 @ 1500000 baud rate

Flash instruction:
- Write sysupgrade image to a microSD card using dd
- Insert the card, and power on the board

Notes on MAC address:
The USB Ethernet (LAN) has a stable, unique MAC address.
MAC address for WAN is LAN's address minus 1.

Signed-off-by: Honpan Lung <lunghonpan@gmail.com>

jmlzr commented:
Where did you even find a schematic? Their (english) download page seems to be missing schematics for all R1 boards. But I can confirm from looking at my LTS board that it's a Motorcomm YT8531C.

From having a first look at the changes in this PR and xunlongs adaption for the LTS (see [1]) I think adding LTS support should only consist of:
* A device tree based on the normal R1 Plus with some very minor changes. (btw xulongs device tree for the normal Plus looks significantly different from yours, did you write your own one for this PR?)
* The 3 C files auto-generated from the device tree that uboot in openwrt needs
* The Linux patch adding YT8531C support
* Something to address the switch from DDR4 to DDR3. But im not sure yet how/where this is done but I think it might be [2].

I might try to cobble something together once I find the time. If so I certainly will update you here on my findings.

The DDR3 vs DDR4 thing needs some changes for sure since I tried this PR on my LTS board and uboot failed with the following error:
```
U-Boot TPL 2021.07 (Jan 06 2022 - 21:22:37)
data training error
row errordata training error
DDR4, 333MHz
BW=16 Col=10 Bk=4 BG=2 CS0 Row=17 CS=1 Die BW=16 Size=2048MB
```

[1] https://github.com/orangepi-xunlong/openwrt/commit/4c813602e618644c3960973e3bbfe024bd299369
[2] https://github.com/orangepi-xunlong/openwrt/commit/4c813602e618644c3960973e3bbfe024bd299369#diff-460d5bea3450aca4031ae9ca505197ef758228bb4593e49367957b4f82b6506e

nebulakl commented:
> Where did you even find a schematic?

http://www.orangepi.org/downloadresources/OrangePiR1PlusLTS/2021-11-29/OrangePiR1PlusLTSc0c5d9f798fc103710a.html

> did you write your own one for this PR?

Yes I wrote mine from scratch

> DDR3 vs DDR4 thing needs some changes for sure

You can create your own `rk3328-orangepi-r1-plus-lts-u-boot.dtsi` based on my `rk3328-orangepi-r1-plus-u-boot.dtsi`, and try replacing `#include "rk3328-sdram-ddr4-666.dtsi"` with `#include "rk3328-sdram-lpddr3-1600.dtsi"`

@jmlzr

jmlzr commented:
Thanks for the tips!

I wrote a device tree for the LTS board and Linux seems to build fine with it but I just can't figure out how to generate the dtoc C files for uboot (like the ones in `of-platdata/orangepi-r1-plus-rk3328`). As far as I understood it I need tho run `tools/dtoc/dtoc` on the `.dtb`. Manually compiling the `.dts` with `dtc` seems to not work and the Makefile doesn't compile the device tree and complains about the missing C files. The whole thing seems to be openwrt/rk3328 specific and I can't find any documentation on it.
Could you please tell me how to do this step?


> Yes I wrote mine from scratch

Is there a reason you started from scratch and not included the one for the extremely similar nanopi-r2s? xulong did it that way and it seems a lot simpler, but I am very new to the world of device trees so not sure if this might be bad practice or something.

Oh and I was wondering as well why did you put the USB LAN into it's ow patch (`107-rockchip-rk3328-enable-LAN-port-on-OrangePi-R1-Plus.patch`) and not directly integrate it into the main one (`105-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch`)?

nebulakl commented:
> reason you started from scratch

Last time I checked Xunlong's device tree for the R1 Plus, it was basically copy-pasted from Nano Pi R2S. I make my own so the names are consistent with those in the schematic diagram.

> why did you put the USB LAN into it's ow patch

A year ago the USB3 was not enabled yet in upstream Linux when I was working on this PR. This is now changed and should be updated.

jmlzr commented:
>Last time I checked Xunlong's device tree for the R1 Plus, it was basically copy-pasted from Nano Pi R2S. I make my own so the names are consistent with those in the schematic diagram.

It's not actually copy pasted, they do `include` the (existing) nanopi r2s device tree and then just specify the differences which makes it quite short. I don't see any actual value in having the names in the device tree match the schematic exactly. As long as it's a accurate description of the HW it should be fine, potentially even beneficial to reuse.

On a practical note: I wrote a device tree for the LTS based on yours and what I learned from the xulong repository. But it seems they also integrated UHS-I support in the LTS (strange that they don't advertise that as it's a nice upgrade)  and the voltage switching (UHS-I uses 1.8V) causes me a lot of trouble. Played around with the device tree quite a bit but can't get around u-boot throwing `Card did not respond to voltage select! : -110` when trying to boot from a UHS-I card.

So next I'm going to start another try based on the current master branch and going xulongs route of basing the device tree on the nanopi r2s since that already has UHS support and I kinda prefer that approach anyway.

nebulakl commented:
I encourage you to try whatever works for you, but please be mindful that [there are certain conventions used by this project](https://openwrt.org/docs/guide-developer/defining-firmware-partitions).

hauke commented on package/boot/uboot-rockchip/patches/102-rockchip-rk3328-Add-support-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,519 @@
+ 	rk3328-rock64.dtb \
+ 	rk3328-rock-pi-e.dtb
+--- /dev/null
++++ b/arch/arm/dts/rk3328-orangepi-r1-plus-u-boot.dtsi
```
Where is this file used?

nebulakl replied:
From https://wiki.st.com/stm32mpu/wiki/STM32MP15_U-Boot
> \<Device tree\>-u-boot.dtsi : addition for u-boot automatically included in build process

We tell u-boot to use ddr4 memory for this board here as the RK3328 supports different types of RAM (LPDDR3/DDR3/DDR4).

hauke commented on target/linux/rockchip/patches-5.10/107-rockchip-rk3328-enable-LAN-port-on-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,54 @@
+ 		regulator-max-microvolt = <5000000>;
+ 	};
++
++	vdd_5v_lan: vdd-5v-lan-regulator {
```
Why do you add this as a separate patch and don't include it in the original patch?

nebulakl replied:
I actually don't know how to do this properly. I do intend to send these changes to upstream Linux and U-Boot. So I tried to create patches based on upstream, then create a separate patch for OpenWrt (like the LED tweak).


Is it better to make the patch specifically for OpenWrt  by consolidating the 3 patch files into 1?

hauke commented on target/linux/rockchip/patches-5.10/106-rockchip-rk3328-OpenWrt-LED-tweak-for-OrangePi-R1-Plus.patch:
```diff
@@ -0,0 +1,16 @@
+ 	model = "Xunlong Orange Pi R1 Plus";
+ 	compatible = "xunlong,orangepi-r1-plus", "rockchip,rk3328";
+ 
++	aliases {
```
Why do you add this as a separate patch and don't include it in the original patch?

hauke reviewed (commented):
Do you plan to send your changes also to the upstream Linux and U-Boot project?

imi415 commented:
> Where did you even find a schematic? Their (english) download page seems to be missing schematics for all R1 boards. But I can confirm from looking at my LTS board that it's a Motorcomm YT8531C.
> 
> From having a first look at the changes in this PR and xunlongs adaption for the LTS (see [1]) I think adding LTS support should only consist of:
> 
> * A device tree based on the normal R1 Plus with some very minor changes. (btw xulongs device tree for the normal Plus looks significantly different from yours, did you write your own one for this PR?)
> * The 3 C files auto-generated from the device tree that uboot in openwrt needs
> * The Linux patch adding YT8531C support
> * Something to address the switch from DDR4 to DDR3. But im not sure yet how/where this is done but I think it might be [2].
> 
> I might try to cobble something together once I find the time. If so I certainly will update you here on my findings.
> 
> The DDR3 vs DDR4 thing needs some changes for sure since I tried this PR on my LTS board and uboot failed with the following error:
> 
> ```
> U-Boot TPL 2021.07 (Jan 06 2022 - 21:22:37)
> data training error
> row errordata training error
> DDR4, 333MHz
> BW=16 Col=10 Bk=4 BG=2 CS0 Row=17 CS=1 Die BW=16 Size=2048MB
> ```
> 
> [1] [orangepi-xunlong/openwrt@4c81360](https://github.com/orangepi-xunlong/openwrt/commit/4c813602e618644c3960973e3bbfe024bd299369) [2] [orangepi-xunlong/openwrt@4c81360#diff-460d5bea3450aca4031ae9ca505197ef758228bb4593e49367957b4f82b6506e](https://github.com/orangepi-xunlong/openwrt/commit/4c813602e618644c3960973e3bbfe024bd299369#diff-460d5bea3450aca4031ae9ca505197ef758228bb4593e49367957b4f82b6506e)

The schematics are available from OrangePi's Chinese website ([Resources](http://www.orangepi.cn/downloadresourcescn/)), the major differences between LTS and original R1 Plus are:
* LPDDR3 instead of LPDDR4 (requires different DRAM initialization code for u-boot)
* Different GbE PHY chip (Motorcomm YT8531C, apparently pin-2-pin compatible with Realtek RTL8111F, but drivers are not compatible)
* The SDHCI voltage select regulator (U3007) is present on LTS, which allows UHS operations.

I personally think the LTS board should be considered as a different model instead of a 'variant'.

KanjiMonster commented:
Orange Pi Plus support was added in ab641efe698f4412319fcbcfe6ffde64c929cd97.
